File: libraries/OclIterator.cs
using System;
using System.Collections;

// Requires SystemTypes OCL library class. 

    class OclIteratorResult
    {
        static OclIteratorResult createOclIteratorResult()
        {
            OclIteratorResult result = new OclIteratorResult();
            return result;
        }

        bool done = false;
        object value = null;

        public static OclIteratorResult newOclIteratorResult(object v)
        {
            OclIteratorResult res = createOclIteratorResult();
            res.value = v;
            if (v == null)
            { res.done = true; }
            else
            { res.done = false; }
            return res;
        }

    }


        class OclIterator
    {
        private int position; // internal
        private int markedPosition; // internal
        private Func<int, object> generatorFunction; 

        public ArrayList elements; // internal
        public ArrayList columnNames; 

        public OclIterator()
        {
            position = 0;
            markedPosition = 0; 
            elements = (new ArrayList());
            generatorFunction = null;
            columnNames = (new ArrayList());
        }

        public override string ToString()
        {
            string _res_ = "(OclIterator) ";
            _res_ = _res_ + position + ",";
            _res_ = _res_ + elements;
            return _res_;
        }

        public void setPosition(int position_x) { position = position_x; }

        public void markPosition() { markedPosition = position; }

        public void movePosition(int i)
        { position = i + position; }

        void moveToMarkedPosition()
        { position = markedPosition; } 

        public void setelements(ArrayList elements_x) { elements = elements_x; }

        public void setelements(int _ind, object elements_x) { elements[_ind] = elements_x; }

        public void addelements(object elements_x)
        { elements.Add(elements_x); }

        public void removeelements(object elements_x)
        { elements = SystemTypes.subtract(elements, elements_x); }

        public int getPosition() { return position; }

        public ArrayList getelements() { return elements; }

        public bool isAfterLast()
        { if (position > elements.Count)
           { return true; }
           else
           { return false; }
        }

        public bool isBeforeFirst()
        { if (position <= 0) 
            { return true; } 
            else 
            { return false; }
        }

        public bool hasNext()
        {
            bool result = false;

            if (position >= 0 && position < elements.Count) { result = true; }
            else { result = false; }
            return result;
        }


        public bool hasPrevious()
        {
            bool result = false;

            if (position > 1 && position <= elements.Count + 1) { result = true; }
            else { result = false; }
            return result;
        }


        public int nextIndex()
        { return position + 1; }


        public int previousIndex()
        {
            return position - 1;
        }


        public void moveForward()
        {
            position++;
        }

        public void moveBackward()
        {
            position--;
        }

        public void moveTo(int i)
        {   position = i;
        }

        public void moveToStart()
        { position = 0; }

        public void moveToEnd()
        {
            position = this.getelements().Count + 1;
        }

        public void moveToFirst()
        { position = 1; }

        public void moveToLast()
        {
            position = this.getelements().Count;
        }

        public static OclIterator newOclIterator_Sequence(ArrayList sq)
        {
            OclIterator ot = new OclIterator();
            ot.setelements(sq);
            ot.setPosition(0);
            return ot;
        }


        public static OclIterator newOclIterator_Set(ArrayList st)
        {
            OclIterator ot = new OclIterator();
            ot.setelements(SystemTypes.sort(st));
            ot.setPosition(0);
            return ot;
        }

        public static OclIterator newOclIterator_String(string str)
        {
            OclIterator ot = new OclIterator();
            ot.elements = SystemTypes.split(str, "[ \n\t\r]+");
            ot.position = 0;
            return ot;
        }

        public static OclIterator newOclIterator_String_String(string str, String chrs)
        {
            OclIterator ot = new OclIterator();
            ot.elements = SystemTypes.split(str, "[" + chrs + "]+");
            ot.position = 0;
            return ot;
        }

        public static OclIterator newOclIterator_Function(Func<int,object> f)
        {
            OclIterator ot = new OclIterator();
            ot.setelements(new ArrayList());
            ot.generatorFunction = f; 
            ot.setPosition(0);
            return ot;
        }

        public object getCurrent()
        {
            object result = null;
            if (position < 1 || position > elements.Count)
            { return result; }

            result = ((object)elements[position - 1]);
            return result;
        }


        public void set(object x)
        {
            if (position < 1 || position > elements.Count)
            { return; }
            elements[position - 1] = x;
        }

        public void insert(object x)
        {
            if (position < 1 || position > elements.Count)
            { return; }

            elements.Insert(position - 1, x);
        }


        public void remove()
        {
            if (position < 1 || position > elements.Count)
            { return; }

            elements.RemoveAt(position - 1);
        }

        public OclIteratorResult nextResult()
        {
            if (generatorFunction == null)
            {
                Object v = next();
                return OclIteratorResult.newOclIteratorResult(v);
            }

            object res = generatorFunction(position);
            position++;
            if (position <= elements.Count)
            { set(res); }
            else
            { elements.Add(res); }
            return OclIteratorResult.newOclIteratorResult(res);
        }


        public object next()
        {
            this.moveForward();
            return this.getCurrent();
        }


        public object previous()
        {
            this.moveBackward();
            return this.getCurrent();
        }

        public object at(int i)
        { return elements[i - 1]; }

        public int length()
        { return elements.Count; }

        public void close()
        {
            position = 0;
            markedPosition = 0;
            elements = new ArrayList();
            columnNames = new ArrayList();
        } 

        public int getColumnCount()
        { return columnNames.Count; }

        public String getColumnName(int i)
        {
            if (columnNames.Count >= i & i >= 1)
            { return (String) columnNames[i-1]; }
            return null;
        }

        public object getCurrentFieldByIndex(int i)
        {
            if (columnNames.Count >= i & i >= 1)
            {
                Hashtable mm = (Hashtable)getCurrent();
                String fld = (String)columnNames[i - 1];
                if (mm != null && fld != null)
                { return mm[fld]; }
            }
            return null;
        }

        public void setCurrentFieldByIndex(int i, object v)
        {
            if (columnNames.Count >= i & i >= 1)
            {
                Hashtable mm = (Hashtable)getCurrent();
                String fld = (String)columnNames[i - 1];
                if (mm != null && fld != null)
                { mm[fld] = v; }
            }
        }
    }

File: libraries/OclIterator.swift
import Foundation
import Darwin

class OclIteratorResult { 
  var done : Bool = true
  var value : Any? = nil
  
  static func newOclIteratorResult(v : Any?) -> OclIteratorResult
  { let res = OclIteratorResult()
    if v == nil { 
      res.done = true 
      res.value = v
    } else {
      res.done = false
      res.value = v!
    }
    return res
  } 
}

class OclIterator
{ static var instance : OclIterator? = nil

  var position : Int = 0
  var markedPosition : Int = 0
  var elements : [Any] = []
  var generatorFunction : (Int) -> Any? = { (i : Int) -> Any? in nil }
  var columnNames : [String] = [String]()

  init() { }

  init(copyFrom: OclIterator) {
    self.position = copyFrom.position
    self.markedPosition = copyFrom.markedPosition
    self.generatorFunction = copyFrom.generatorFunction
    self.elements = Ocl.copySequence(s: copyFrom.elements)
    self.columnNames = Ocl.copySequence(s: copyFrom.columnNames)
  }

  func copy() -> OclIterator
  { let res : OclIterator = OclIterator(copyFrom: self)
    OclIterator_allInstances.append(res)
    return res
  }

  static func defaultInstance() -> OclIterator
  { if (instance == nil)
    { instance = createOclIterator() }
    return instance!
  }


  func hasNext() -> Bool
  {
    var result : Bool = false
    if position >= 0 && position < elements.count
    {
      result = true
    }
    else {
      result = false
    }
    return result

  }


  func hasPrevious() -> Bool
  {
    var result : Bool = false
    if position > 1 && position <= elements.count + 1
    {
      result = true
    }
    else {
      result = false
    }
    return result

  }

  func isAfterLast() -> Bool
  {
    if position > elements.count
    {
      return true
    }
    return false
  }

  func isBeforeFirst() -> Bool
  {
    if position <= 0
    {
      return true
    }
    return false
  }

  func nextIndex() -> Int
  {
    var result : Int = 0
    result = position + 1
    return result
  }


  func previousIndex() -> Int
  {
    var result : Int = 0
    result = position - 1
    return result
  }


  func moveForward() -> Void
  {
    position = position + 1
  }


  func moveBackward() -> Void
  {
    position = position - 1
  }


  func moveTo(i : Int) -> Void
  {
    position = i
  }

  func moveToFirst()
  {
    position = 1
  }

  func moveToLast()
  {
    position = elements.count
  }
  
  func moveToStart()
  {
    position = 0
  }

  func moveToEnd()
  {
    position = elements.count + 1
  }
  
  func getPosition() -> Int
  { return position }
  
  func setPosition(i : Int)
  { position = i }
 
  func markPosition()
  { markedPosition = position }

  func moveToMarkedPosition()
  { position = markedPosition }
  
  func movePosition(i : Int)
  { position = position + i } 

  static func newOclIterator_Sequence(sq : [Any]) -> OclIterator
  {
    let ot : OclIterator = createOclIterator()
    ot.elements = sq
    ot.position = 0
    ot.markedPosition = 0
    return ot
  }


  static func newOclIterator_Set<T : Hashable>(st : Set<T>) -> OclIterator
  {
    let ot : OclIterator = createOclIterator()
    ot.elements = Ocl.toSequence(s: st)
    ot.position = 0
    ot.markedPosition = 0
    return ot
  }

  static func newOclIterator_String(ss : String) -> OclIterator
  { let ot : OclIterator = createOclIterator()
    ot.elements = Ocl.split(str: ss, pattern: "[ \n\t\r]+")
    ot.position = 0
    ot.markedPosition = 0
    return ot
  }


  static func newOclIterator_String_String(ss : String, seps : String) -> OclIterator
  { let ot : OclIterator = createOclIterator()
    ot.elements = Ocl.split(str: ss, pattern: "[" + seps + "]+")
    ot.position = 0
    ot.markedPosition = 0
    return ot
  }

  static func newOclIterator_Function(f : @escaping (Int) -> Any?) -> OclIterator
  {
    let ot : OclIterator = createOclIterator()
    ot.elements = []
    ot.position = 0
    ot.markedPosition = 0
    ot.generatorFunction = f
    return ot
  }

  func getCurrent() -> Any
  {
    var result : Any? = nil
    if position > 0 && position <= elements.count 
    { result = elements[position - 1]
      return result!
    } 
    return Double.nan
  }


  func set(x : Any) -> Void
  {
    elements[position-1] = x
  }


  func insert(x : Any) -> Void
  { elements.insert(x, at: position-1) }


  func remove() -> Void
  { elements.remove(at: position) }

  func nextResult() -> OclIteratorResult
  { let v = generatorFunction(position)
    position = position + 1
    if v == nil 
    { return OclIteratorResult.newOclIteratorResult(v: v) } 
    
    if position <= elements.count 
    { set(x: v!) } 
    else
    { elements.append(v!) }
      
    return OclIteratorResult.newOclIteratorResult(v: v)
  } 

  func next() -> Any
  {
    moveForward()
    return getCurrent()
  } 
  /* Signal end of iteration with Double.nan */ 


  func previous() -> Any
  {
    moveBackward()
    return getCurrent()
  }

  func at(i : Int) -> Any
  {
    var result : Any? = nil
    result = elements[i - 1]
    return result!
  }

  func length() -> Int
  { return elements.count }

  func getColumnCount() -> Int
  { return columnNames.count }

  func getColumnName(i : Int) -> String
  { return columnNames[i-1] } 

  func getCurrentFieldByIndex(i : Int) -> Any
  { let m = getCurrent() as! [String:Any]
    return m[columnNames[i-1]]!
  }  

  func setCurrentFieldByIndex(i : Int, v : Any)
  { var m = getCurrent() as! [String:Any]
    m[columnNames[i-1]] = v
  }

  func close()
  { elements = []
    position = 0
    columnNames = []
    markedPosition = 0
  } 
}

var OclIterator_allInstances : [OclIterator] = [OclIterator]()

func createOclIterator() -> OclIterator
{ let result : OclIterator = OclIterator()
  OclIterator_allInstances.append(result)
  return result
}

func killOclIterator(obj: OclIterator)
{ OclIterator_allInstances = OclIterator_allInstances.filter{ $0 !== obj } }


